src/clashing_inputs.rs

```diff
@@ -47,36 +46,87 @@ impl ClashStrategy {
     }
 }
 
-impl UserInput {
-    /// Does `self` clash with `other`?
-    #[must_use]
-    fn clashes(&self, other: &UserInput) -> bool {
-        use UserInput::*;
+/// The basic inputs that make up a [`UserInput`].
+#[derive(Debug, Clone)]
+#[must_use]
+pub enum BasicInputs {
+    /// The input consists of a single, fundamental [`UserInput`].
+    /// In most cases, the input simply holds itself.
+    Simple(Box<dyn UserInput>),
+
+    /// The input consists of multiple independent [`UserInput`]s.
+    Composite(Vec<Box<dyn UserInput>>),
 
+    /// The input represents one or more independent [`UserInput`] types.
+    Group(Vec<Box<dyn UserInput>>),
+}
+
+impl BasicInputs {
+    /// Returns a list of the underlying [`UserInput`]s.
+    #[inline]
+    pub fn inputs(&self) -> Vec<Box<dyn UserInput>> {
+        match self.clone() {
+            Self::Simple(input) => vec![input],
+            Self::Composite(inputs) => inputs,
+            Self::Group(inputs) => inputs,
+        }
+    }
+
+    /// Returns the number of the logical [`UserInput`]s that make up the input.
+    #[allow(clippy::len_without_is_empty)]
+    #[inline]
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Simple(_) => 1,
+            Self::Composite(_) => 1,
```

Comment:
It's quite surprising that this is 1. If it's not a bug, it needs a doc comment.

Reply:
This is the old `UserInput::len`